### DISCUSSION TITLE
Will Pocketbase work as a sync server?

### DISCUSSION THREAD

**Discussion by fbede** [pocketbase/pocketbase#3337] [Q&A] [answered]
I have tried figuring this out on my own but to no avail. My planned use case for pocketbase is a sync server for an offline first application. Due to the nature of my use case, the server would be getting large infrequent write requests when the client has been offline for a while and connects back to the server. As the number of users increases, these large write requests would only get more frequent. From looking at the benchmarks, the performance of writes did not seem to improve when I compare the results of the 2vcpu Hetzer server to the 16vcpu one. 

I am just wondering if my concerns are legitimate or I am worried for nothing.

If I am worried for nothing how many users would be too many for this use case.

Here's the [link](https://github.com/pocketbase/benchmarks) to the benchmark if you have not seen it.

**Comment by ganigeorgiev** [comment, marked as answer]
> From looking at the benchmarks, the performance of writes did not seem to improve when I compare the results of the 2vcpu Hetzer server to the 16vcpu one.

As mentioned in the ["Takeaways" section of the benchmarks](https://github.com/pocketbase/benchmarks#takeaways-and-things-well-have-to-improve) this is probably because the operations are IO bound, aka. they depend mostly on the disk speed.
Having a more beefy hardware may not necessary help with the individual query execution but it will definitely help with the more heavy concurrent scenarios.

> ...the server would be getting large infrequent write requests when the client has been offline ...
> If I am worried for nothing how many users would be too many for this use case.

This would be hard to answer as it really depends on the specific load.
My recommendation would be to create a small POC and try to simulate your expected usage scenario.
Additionally, instead of sending individual HTTP write requests, it would be more efficient for your specific use case to create a custom route that will accept bulk/batch submission.

Keep in mind that you can have only 1 writer at a time (other writes queue up). From the [SQLite docs](https://www.sqlite.org/whentouse.html):

> SQLite only supports one writer at a time per database file. But in most cases, a write transaction only takes milliseconds and so multiple writers can simply take turns. SQLite will handle more write concurrency than many people suspect. Nevertheless, client/server database systems, because they have a long-running server process at hand to coordinate access, can usually handle far more write concurrency than SQLite ever will.

_For other general SQLite limits you can also check https://www.sqlite.org/limits.html._
